File: SL-Back-end/app/schemas/community.py
"""
커뮤니티 API 스키마
- 게시글, 댓글, 좋아요 관련 Request/Response 스키마
"""
from pydantic import BaseModel, Field, field_serializer
from typing import Optional, List
from datetime import datetime
from decimal import Decimal


# ============================================================
# 게시글 관련 스키마
# ============================================================

class PostCreate(BaseModel):
    """게시글 작성 요청"""
    title: str = Field(..., min_length=1, max_length=200, description="제목")
    content: str = Field(..., min_length=1, description="내용")
    tags: Optional[List[str]] = Field(None, description="태그 목록")
    post_type: str = Field("DISCUSSION", description="게시글 유형 (STRATEGY_SHARE/DISCUSSION/QUESTION)")

    # 전략 공유인 경우
    strategy_id: Optional[str] = Field(None, description="공유할 전략 ID")
    session_id: Optional[str] = Field(None, description="공유할 백테스트 세션 ID")
    is_anonymous: bool = Field(False, description="익명 여부")


class PostUpdate(BaseModel):
    """게시글 수정 요청"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    content: Optional[str] = Field(None, min_length=1)
    tags: Optional[List[str]] = None


class PostSummary(BaseModel):
    """게시글 목록용 요약 정보"""
    post_id: str = Field(..., serialization_alias="postId")
    title: str
    content_preview: str = Field(..., serialization_alias="contentPreview", description="내용 미리보기 (한 줄)")
    author_nickname: Optional[str] = Field(None, serialization_alias="authorNickname")
    author_id: Optional[str] = Field(None, serialization_alias="authorId", description="작성자 ID")
    is_anonymous: bool = Field(..., serialization_alias="isAnonymous")
    tags: Optional[List[str]] = None
    post_type: str = Field(..., serialization_alias="postType")
    session_snapshot: Optional['SessionSnapshot'] = Field(
        None, serialization_alias="sessionSnapshot"
    )

    # 통계
    view_count: int = Field(..., serialization_alias="viewCount")
    like_count: int = Field(..., serialization_alias="likeCount")
    comment_count: int = Field(..., serialization_alias="commentCount")

    # 시간
    created_at: datetime = Field(..., serialization_alias="createdAt")
    updated_at: datetime = Field(..., serialization_alias="updatedAt")

    class Config:
        populate_by_name = True


class StrategySnapshot(BaseModel):
    """전략 스냅샷 (공유용)"""
    strategy_name: str = Field(..., serialization_alias="strategyName")
    strategy_type: Optional[str] = Field(None, serialization_alias="strategyType")
    description: Optional[str] = None
    buy_conditions: List[dict] = Field(..., serialization_alias="buyConditions")
    sell_conditions: dict = Field(..., serialization_alias="sellConditions")
    trade_targets: dict = Field(..., serialization_alias="tradeTargets")

    class Config:
        populate_by_name = True


class SessionSnapshot(BaseModel):
    """백테스트 세션 스냅샷 (공유용)"""
    initial_capital: Decimal = Field(..., serialization_alias="initialCapital")
    start_date: str = Field(..., serialization_alias="startDate")
    end_date: str = Field(..., serialization_alias="endDate")
    total_return: Decimal = Field(..., serialization_alias="totalReturn")
    annualized_return: Optional[Decimal] = Field(None, serialization_alias="annualizedReturn")
    max_drawdown: Optional[Decimal] = Field(None, serialization_alias="maxDrawdown")
    sharpe_ratio: Optional[Decimal] = Field(None, serialization_alias="sharpeRatio")
    win_rate: Optional[Decimal] = Field(None, serialization_alias="winRate")

    @field_serializer('initial_capital', 'total_return', 'annualized_return', 'max_drawdown', 'sharpe_ratio', 'win_rate')
    def serialize_decimal(self, value: Optional[Decimal]) -> Optional[float]:
        return float(value) if value is not None else None

    class Config:
        populate_by_name = True


class PostDetail(BaseModel):
    """게시글 상세 정보"""
    post_id: str = Field(..., serialization_alias="postId")
    title: str
    content: str
    author_nickname: Optional[str] = Field(None, serialization_alias="authorNickname")
    author_id: Optional[str] = Field(None, serialization_alias="authorId", description="작성자 ID (본인 여부 확인용)")
    is_anonymous: bool = Field(..., serialization_alias="isAnonymous")
    tags: Optional[List[str]] = None
    post_type: str = Field(..., serialization_alias="postType")

    # 전략 공유 정보
    strategy_snapshot: Optional[StrategySnapshot] = Field(None, serialization_alias="strategySnapshot")
    session_snapshot: Optional[SessionSnapshot] = Field(None, serialization_alias="sessionSnapshot")

    # 통계
    view_count: int = Field(..., serialization_alias="viewCount")
    like_count: int = Field(..., serialization_alias="likeCount")
    comment_count: int = Field(..., serialization_alias="commentCount")
    is_liked: bool = Field(False, serialization_alias="isLiked", description="현재 사용자의 좋아요 여부")

    # 시간
    created_at: datetime = Field(..., serialization_alias="createdAt")
    updated_at: datetime = Field(..., serialization_alias="updatedAt")

    class Config:
        populate_by_name = True


class PostListResponse(BaseModel):
    """게시글 목록 응답"""
    posts: List[PostSummary]
    total: int
    page: int
    limit: int
    has_next: bool = Field(..., serialization_alias="hasNext")

    class Config:
        populate_by_name = True


# ============================================================
# 댓글 관련 스키마
# ============================================================

class CommentCreate(BaseModel):
    """댓글 작성 요청"""
    content: str = Field(..., min_length=1, description="댓글 내용")
    parent_comment_id: Optional[str] = Field(None, serialization_alias="parentCommentId", description="부모 댓글 ID (대댓글인 경우)")
    is_anonymous: bool = Field(False, serialization_alias="isAnonymous")

    class Config:
        populate_by_name = True


class CommentUpdate(BaseModel):
    """댓글 수정 요청"""
    content: str = Field(..., min_length=1)


class CommentItem(BaseModel):
    """댓글 정보"""
    comment_id: str = Field(..., serialization_alias="commentId")
    post_id: str = Field(..., serialization_alias="postId")
    content: str
    author_nickname: Optional[str] = Field(None, serialization_alias="authorNickname")
    author_id: Optional[str] = Field(None, serialization_alias="authorId")
    is_anonymous: bool = Field(..., serialization_alias="isAnonymous")
    parent_comment_id: Optional[str] = Field(None, serialization_alias="parentCommentId")
    like_count: int = Field(..., serialization_alias="likeCount")
    is_liked: bool = Field(False, serialization_alias="isLiked")
    created_at: datetime = Field(..., serialization_alias="createdAt")
    updated_at: datetime = Field(..., serialization_alias="updatedAt")

    # 대댓글 목록 (재귀)
    replies: List['CommentItem'] = []

    class Config:
        populate_by_name = True


class CommentListResponse(BaseModel):
    """댓글 목록 응답"""
    comments: List[CommentItem]
    total: int


# ============================================================
# 좋아요 관련 스키마
# ============================================================

class LikeResponse(BaseModel):
    """좋아요 응답"""
    is_liked: bool = Field(..., serialization_alias="isLiked", description="좋아요 상태")
    like_count: int = Field(..., serialization_alias="likeCount", description="총 좋아요 수")

    class Config:
        populate_by_name = True


# ============================================================
# 랭킹 관련 스키마
# ============================================================

class RankingItem(BaseModel):
    """수익률 랭킹 항목"""
    rank: int = Field(..., description="순위")
    strategy_id: str = Field(..., serialization_alias="strategyId")
    session_id: str = Field(..., serialization_alias="sessionId")
    strategy_name: str = Field(..., serialization_alias="strategyName")
    author_nickname: Optional[str] = Field(None, serialization_alias="authorNickname")
    total_return: Decimal = Field(..., serialization_alias="totalReturn")
    annualized_return: Optional[Decimal] = Field(None, serialization_alias="annualizedReturn")
    max_drawdown: Optional[Decimal] = Field(None, serialization_alias="maxDrawdown")
    sharpe_ratio: Optional[Decimal] = Field(None, serialization_alias="sharpeRatio")

    @field_serializer('total_return', 'annualized_return', 'max_drawdown', 'sharpe_ratio')
    def serialize_decimal(self, value: Optional[Decimal]) -> Optional[float]:
        return float(value) if value is not None else None

    class Config:
        populate_by_name = True


class TopRankingsResponse(BaseModel):
    """상위 랭킹 응답 (상위 3개)"""
    rankings: List[RankingItem]
    updated_at: datetime = Field(..., serialization_alias="updatedAt", description="마지막 업데이트 시간")

    class Config:
        populate_by_name = True

# ForwardRef 해제
PostSummary.model_rebuild()


class RankingListResponse(BaseModel):
    """전체 랭킹 목록 응답"""
    rankings: List[RankingItem]
    total: int
    page: int
    limit: int

    class Config:
        populate_by_name = True


# ============================================================
# 전략 복제 관련 스키마
# ============================================================

class CloneStrategyData(BaseModel):
    """전략 복제 데이터"""
    # 기본 설정
    strategy_name: str = Field(..., serialization_alias="strategyName")
    is_day_or_month: str = Field(..., serialization_alias="isDayOrMonth")
    initial_investment: int = Field(..., serialization_alias="initialInvestment")
    start_date: str = Field(..., serialization_alias="startDate")
    end_date: str = Field(..., serialization_alias="endDate")
    commission_rate: float = Field(..., serialization_alias="commissionRate")
    slippage: float = Field(..., serialization_alias="slippage")

    # 매수 조건
    buy_conditions: List[dict] = Field(..., serialization_alias="buyConditions")
    buy_logic: str = Field(..., serialization_alias="buyLogic")
    priority_factor: Optional[str] = Field(None, serialization_alias="priorityFactor")
    priority_order: str = Field(..., serialization_alias="priorityOrder")

    # 매수 비중
    per_stock_ratio: float = Field(..., serialization_alias="perStockRatio")
    max_holdings: int = Field(..., serialization_alias="maxHoldings")
    max_buy_value: Optional[int] = Field(None, serialization_alias="maxBuyValue")
    max_daily_stock: Optional[int] = Field(None, serialization_alias="maxDailyStock")
    buy_price_basis: str = Field(..., serialization_alias="buyPriceBasis")
    buy_price_offset: float = Field(..., serialization_alias="buyPriceOffset")

    # 매도 조건
    target_and_loss: Optional[dict] = Field(None, serialization_alias="targetAndLoss")
    hold_days: Optional[dict] = Field(None, serialization_alias="holdDays")
    condition_sell: Optional[dict] = Field(None, serialization_alias="conditionSell")

    # 매매 대상
    trade_targets: dict = Field(..., serialization_alias="tradeTargets")

    class Config:
        populate_by_name = True

File: SL-Front-End/src/lib/api/community.ts
import { axiosInstance } from "../axios";

// ============================================================
// 타입 정의
// ============================================================

export interface PostSummary {
  postId: string;
  title: string;
  contentPreview: string;
  authorNickname: string | null;
  authorId: string | null;
  isAnonymous: boolean;
  tags: string[] | null;
  postType: string;
  viewCount: number;
  likeCount: number;
  commentCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface StrategySnapshot {
  strategyName: string;
  strategyType: string | null;
  description: string | null;
  buyConditions: Record<string, unknown>[];
  sellConditions: Record<string, unknown>;
  tradeTargets: Record<string, unknown>;
}

export interface SessionSnapshot {
  initialCapital: number;
  startDate: string;
  endDate: string;
  totalReturn: number;
  annualizedReturn: number | null;
  maxDrawdown: number | null;
  sharpeRatio: number | null;
  winRate: number | null;
}

export interface PostDetail {
  postId: string;
  title: string;
  content: string;
  authorNickname: string | null;
  authorId: string | null;
  isAnonymous: boolean;
  tags: string[] | null;
  postType: string;
  strategySnapshot: StrategySnapshot | null;
  sessionSnapshot: SessionSnapshot | null;
  viewCount: number;
  likeCount: number;
  commentCount: number;
  isLiked: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface PostListResponse {
  posts: PostSummary[];
  total: number;
  page: number;
  limit: number;
  hasNext: boolean;
}

export interface PostCreate {
  title: string;
  content: string;
  tags?: string[];
  postType?: string;
  strategyId?: string;
  sessionId?: string;
  isAnonymous?: boolean;
}

export interface PostUpdate {
  title?: string;
  content?: string;
  tags?: string[];
}

export interface CommentItem {
  commentId: string;
  postId: string;
  content: string;
  authorNickname: string | null;
  authorId: string | null;
  isAnonymous: boolean;
  parentCommentId: string | null;
  likeCount: number;
  isLiked: boolean;
  createdAt: string;
  updatedAt: string;
  replies: CommentItem[];
}

export interface CommentListResponse {
  comments: CommentItem[];
  total: number;
}

export interface CommentCreate {
  content: string;
  parentCommentId?: string;
  isAnonymous?: boolean;
}

export interface CommentUpdate {
  content: string;
}

export interface LikeResponse {
  isLiked: boolean;
  likeCount: number;
}

export interface RankingItem {
  rank: number;
  strategyId: string;
  sessionId: string;
  strategyName: string;
  authorNickname: string | null;
  totalReturn: number;
  annualizedReturn: number | null;
  maxDrawdown: number | null;
  sharpeRatio: number | null;
}

export interface TopRankingsResponse {
  rankings: RankingItem[];
  updatedAt: string;
}

export interface RankingListResponse {
  rankings: RankingItem[];
  total: number;
  page: number;
  limit: number;
}

export interface CloneStrategyData {
  strategyName: string;
  isDayOrMonth: string;
  initialInvestment: number;
  startDate: string;
  endDate: string;
  commissionRate: number;
  slippage: number;
  buyConditions: Record<string, unknown>[];
  buyLogic: string;
  priorityFactor: string | null;
  priorityOrder: string;
  perStockRatio: number;
  maxHoldings: number;
  maxBuyValue: number | null;
  maxDailyStock: number | null;
  buyPriceBasis: string;
  buyPriceOffset: number;
  targetAndLoss: Record<string, unknown> | null;
  holdDays: Record<string, unknown> | null;
  conditionSell: Record<string, unknown> | null;
  tradeTargets: Record<string, unknown>;
}

// ============================================================
// 게시글 API
// ============================================================

export const communityApi = {
  /**
   * 게시글 목록 조회
   */
  getPosts: async (params?: {
    postType?: string;
    tags?: string | string[];
    search?: string;
    userId?: string;
    page?: number;
    limit?: number;
  }): Promise<PostListResponse> => {
    const { postType, userId, tags, search, page, limit } = params || {};

    // API가 기대하는 형태로 매핑
    const tagsParam =
      tags === undefined || tags === null || tags === ""
        ? undefined
        : Array.isArray(tags)
          ? tags.join(",")
          : tags;

    const response = await axiosInstance.get<PostListResponse>(
      "/community/posts",
      {
        params: {
          post_type: postType,
          user_id: userId, // camelCase를 snake_case로 변환
          tags: tagsParam,
          search,
          page,
          limit,
        },
      },
    );
    return response.data;
  },

  /**
   * 게시글 상세 조회
   */
  getPost: async (postId: string): Promise<PostDetail> => {
    const response = await axiosInstance.get<PostDetail>(
      `/community/posts/${postId}`
    );
    return response.data;
  },

  /**
   * 게시글 작성
   */
  createPost: async (data: PostCreate): Promise<PostDetail> => {
    const response = await axiosInstance.post<PostDetail>(
      "/community/posts",
      data
    );
    return response.data;
  },

  /**
   * 게시글 수정
   */
  updatePost: async (
    postId: string,
    data: PostUpdate
  ): Promise<PostDetail> => {
    const response = await axiosInstance.put<PostDetail>(
      `/community/posts/${postId}`,
      data
    );
    return response.data;
  },

  /**
   * 게시글 삭제
   */
  deletePost: async (postId: string): Promise<void> => {
    await axiosInstance.delete(`/community/posts/${postId}`);
  },

  // ============================================================
  // 댓글 API
  // ============================================================

  /**
   * 댓글 목록 조회 (트리 구조)
   */
  getComments: async (postId: string): Promise<CommentListResponse> => {
    const response = await axiosInstance.get<CommentListResponse>(
      `/community/posts/${postId}/comments`
    );
    return response.data;
  },

  /**
   * 댓글 작성
   */
  createComment: async (
    postId: string,
    data: CommentCreate
  ): Promise<CommentItem> => {
    const response = await axiosInstance.post<CommentItem>(
      `/community/posts/${postId}/comments`,
      data
    );
    return response.data;
  },

  /**
   * 댓글 수정
   */
  updateComment: async (
    commentId: string,
    data: CommentUpdate
  ): Promise<CommentItem> => {
    const response = await axiosInstance.put<CommentItem>(
      `/community/comments/${commentId}`,
      data
    );
    return response.data;
  },

  /**
   * 댓글 삭제
   */
  deleteComment: async (commentId: string): Promise<void> => {
    await axiosInstance.delete(`/community/comments/${commentId}`);
  },

  // ============================================================
  // 좋아요 API
  // ============================================================

  /**
   * 게시글 좋아요 토글
   */
  togglePostLike: async (postId: string): Promise<LikeResponse> => {
    const response = await axiosInstance.post<LikeResponse>(
      `/community/posts/${postId}/like`
    );
    return response.data;
  },

  /**
   * 댓글 좋아요 토글
   */
  toggleCommentLike: async (commentId: string): Promise<LikeResponse> => {
    const response = await axiosInstance.post<LikeResponse>(
      `/community/comments/${commentId}/like`
    );
    return response.data;
  },

  // ============================================================
  // 랭킹 API
  // ============================================================

  /**
   * 상위 3개 랭킹 조회
   */
  getTopRankings: async (): Promise<TopRankingsResponse> => {
    const response = await axiosInstance.get<TopRankingsResponse>(
      "/community/rankings/top"
    );
    return response.data;
  },

  /**
   * 전체 랭킹 목록 조회
   */
  getRankings: async (params?: {
    page?: number;
    limit?: number;
  }): Promise<RankingListResponse> => {
    const response = await axiosInstance.get<RankingListResponse>(
      "/community/rankings",
      { params }
    );
    return response.data;
  },

  // ============================================================
  // 전략 복제 API
  // ============================================================

  /**
   * 복제용 전략 데이터 조회
   */
  getCloneStrategyData: async (
    sessionId: string
  ): Promise<CloneStrategyData> => {
    const response = await axiosInstance.get<CloneStrategyData>(
      `/community/clone-strategy/${sessionId}`
    );
    return response.data;
  },

  /**
   * 전략 복제 실행
   */
  cloneStrategy: async (sessionId: string): Promise<{ message: string }> => {
    const response = await axiosInstance.post<{ message: string }>(
      `/community/clone-strategy/${sessionId}`
    );
    return response.data;
  },
};
